frameworks/Java/play2-java/play2-java-jooq-hikaricp/app/controllers/Application.java

```diff
@@ -0,0 +1,121 @@
+    private static final JSONFormat JSON_FORMAT = JSONFormat.DEFAULT_FOR_RECORDS.recordFormat(RecordFormat.OBJECT);
+
+    private final Database db;
+    private final DatabaseExecutionContext dbEc;
+
+    @Inject
+    public Application(final Database db, final DatabaseExecutionContext dbEc) {
+        this.db = db;
+        this.dbEc = dbEc;
+    }
+
+    public CompletionStage<Result> db() {
+        return getRandomWorlds(1).thenApply(worlds -> ok(worlds.get(0).formatJSON(JSON_FORMAT)).as(JSON));
+    }
+
+    public CompletionStage<Result> queries(final String queries) {
+        return getRandomWorlds(queryCount(queries)).thenApply(worlds -> ok(worlds.formatJSON(JSON_FORMAT)).as(JSON));
+    }
+
+    public CompletionStage<Result> fortunes() {
+        return CompletableFuture.supplyAsync(() -> {
+            final List<FortuneRecord> fortunes = this.db.withConnection(connection -> {
+                return DSL.using(connection, DIALECT).selectFrom(FORTUNE).fetch();
+            });
+            fortunes.add(new FortuneRecord(UInteger.valueOf(0), "Additional fortune added at request time."));
+            Collections.sort(fortunes, (f1, f2) -> f1.getMessage().compareTo(f2.getMessage()));
+
+            return ok(views.html.fortunes.render(fortunes));
+        }, dbEc);
+    }
+
+    public CompletionStage<Result> update(final String queries) {
+        return getRandomWorlds(queryCount(queries)).thenApplyAsync(worlds -> {
+            final Random random = ThreadLocalRandom.current();
+            for (final WorldRecord world : worlds) {
+                world.setRandomnumber((random.nextInt(10000) + 1));
+            }
+
+            final int batchSize = 25;
+            final int batches = ((worlds.size() / batchSize) + 1);
+            this.db.withConnection(connection -> {
+                final DSLContext sql = DSL.using(connection, DIALECT);
+                for ( int i = 0 ; i < batches ; ++i ) {
+                    sql.batchUpdate(worlds.subList(i * batchSize, Math.min((i + 1) * batchSize, worlds.size()))).execute();
+                }
+                return null;
+            });
+
+            return ok(worlds.formatJSON(JSON_FORMAT)).as(JSON);
+        }, dbEc);
+    }
+
+    private int queryCount(final String queryCountString) {
+        int queryCount;
+        try {
+            queryCount = Integer.parseInt(queryCountString, 10);
+        } catch (final NumberFormatException e) {
+            queryCount = 1;
+        }
+        if (queryCount < 1) {
+            queryCount = 1;
+        } else if (queryCount > 500) {
+            queryCount = 500;
+        }
+
+        return queryCount;
+    }
+
+    private CompletionStage<org.jooq.Result<WorldRecord>> getRandomWorlds(final int n) {
+        return CompletableFuture.supplyAsync(() -> {
+            final Random random = ThreadLocalRandom.current();
+            org.jooq.Result<WorldRecord> worlds = null;
+            for (int i = 0; i < n; ++i) {
```

Comment:
If `n` is large, as a SQL person, I might be inclined to create a temp table, insert all the random values in that, and then semi-join that table to the `WORLD` table to get all the rows in one go. Given how this algorithm is implemented here, this is probably a more reasonable approach.

Alternatively, if this test is run against H2, HSQLDB, Oracle, PostgreSQL, or SQL Server, an array parameter or table-valued parameter might be suitable as well for use in an `IN` predicate.

I think the jOOQ approach would favour something like this.

Reply:
`IN` is not allowed, please read the test requirements.

Reply:
Yeah, I've seen it now, but *why not*? If a benchmark coerces a framework into non-idiomatic usage, then the benchmark is pretty meaningless, in my opinion.

Reply:
You should talk to the @TechEmpower people to discuss the *why* :smile:

Reply:
You're right of course!